### PR TITLE
[WIP] Fix relationship with user using microservice and oauth

### LINKED ISSUE #8550: Microservices with oauth2 auth and ManyToOne relationship to User with services fail to compile.
##### **Overview of the issue**

Microservices with oauth2 auth and ManyToOne relationship to User with services fail to compile.

Given the following JDL:

```
application {
  config {
    baseName gateway,
    applicationType gateway,
    authenticationType oauth2,
    packageName com.example.gateway,
    prodDatabaseType postgresql,
    clientFramework react,
    buildTool gradle
  }
  entities A, B, C, D
}

application {
  config {
    baseName microservicea,
    applicationType microservice,
    authenticationType oauth2,
    packageName com.example.microservicea,
    prodDatabaseType postgresql,
    buildTool gradle
  }
  entities A, B
}

application {
  config {
    baseName microserviceb,
    applicationType microservice,
    authenticationType oauth2,
    packageName com.example.microserviceb,
    prodDatabaseType postgresql,
    buildTool gradle
  }
  entities C, D
}

entity A {
    data String
}

entity B {
    data String
}

entity C {
    data String
}

entity D {
    data String
}

relationship ManyToOne {
  A{user(login)} to User
}

dto * with mapstruct
service * with serviceClass

microservice A, B with microservicea
microservice C, D with microserviceb
```

building microservicea with `./gradlew bootWar -Pprod jibDockerBuild`:

```
> Task :compileJava
Note: Hibernate JPA 2 Static-Metamodel Generator 5.2.17.Final
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/service/AService.java:31: error: cannot find symbol
    private final UserRepository userRepository;
                  ^
  symbol:   class UserRepository
  location: class AService
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/service/AService.java:33: error: cannot find symbol
    public AService(ARepository aRepository, AMapper aMapper, UserRepository userRepository) {
                                                              ^
  symbol:   class UserRepository
  location: class AService
2 errors

> Task :compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 4s
4 actionable tasks: 4 executed
```

If `dto * with mapstruct` and `service * with serviceClass` are commented out in the jdl, the microservice will build correctly.

##### **Motivation for or Use Case**

Discovered while trying to do filtering on an entity, which requires a service. 

Tried the okta microservices demo at: https://github.com/oktadeveloper/okta-jhipster-microservices-oauth-example

Realized that example does not use services.

The documentation at Using DTO's (https://www.jhipster.tech/using-dtos) reads that this should be valid.

It may be worth noting that changing the auth type to JWT fails compilation as well:

```
> Task :compileJava FAILED
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/domain/A.java:32: error: cannot find symbol
    private User user;
            ^
  symbol:   class User
  location: class A
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/domain/A.java:56: error: cannot find symbol
    public User getUser() {
           ^
  symbol:   class User
  location: class A
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/domain/A.java:60: error: cannot find symbol
    public A user(User user) {
                  ^
  symbol:   class User
  location: class A
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/domain/A.java:65: error: cannot find symbol
    public void setUser(User user) {
                        ^
  symbol:   class User
  location: class A
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/service/mapper/AMapper.java:11: error: cannot find symbol
@Mapper(componentModel = "spring", uses = {UserMapper.class})
                                           ^
  symbol: class UserMapper
/Users/jsm174/git/jhipster/react-test-9/microservicea/src/main/java/com/example/microservicea/service/mapper/AMapper.java:11: error: Couldn't retrieve @Mapper annotation
@Mapper(componentModel = "spring", uses = {UserMapper.class})
^
Note: Hibernate JPA 2 Static-Metamodel Generator 5.2.17.Final
6 errors
```

 

##### **JHipster Version(s)**

```
microservicea@0.0.0 /Users/jsm174/git/jhipster/react-test-9/microservicea
└── generator-jhipster@5.4.2

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "clientPackageManager": "npm",
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": [
      "en",
      "fr"
    ],
    "messageBroker": false,
    "nativeLanguage": "en",
    "packageName": "com.example.microservicea",
    "packageFolder": "com/example/microservicea",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "serviceDiscoveryType": "eureka",
    "skipClient": true,
    "testFrameworks": [],
    "websocket": false,
    "baseName": "microservicea",
    "applicationType": "microservice",
    "authenticationType": "oauth2",
    "buildTool": "gradle",
    "jhipsterVersion": "5.4.2",
    "skipUserManagement": true,
    "cacheProvider": "hazelcast",
    "serverPort": "8081"
  },
  "entities": [
    "A",
    "B"
  ]
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity A {
  data String
}
entity B {
  data String
}
relationship ManyToOne {
  A{user(login)} to User
}

dto A, B with mapstruct
service A, B with serviceClass
microservice A, B with microservicea

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)

git version 2.17.1 (Apple Git-112)

node: v10.11.0

npm: 6.4.1

yarn: 1.10.1

Docker version 18.06.1-ce, build e68fc7a

docker-compose version 1.22.0, build f46880f

-   [X] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by Falydoor**
Fix #8550 

I did a refactor of inject_template to avoid having repeated conditions and have something more "clear".

The issue was because when having an entity using DTO with a relationship with User the resource can't save the User since the DTO doesn't contain all user's data.

@mraible do you think the logic below will need to be the same in case of an entity using DTO? If yes that means the DTO must contain all user's data...
```javascript
<%_ if (saveUserSnapshot) { %>
if (<%= instanceName %>.getUser() != null) {
    // Save user in case it's new and only exists in gateway
    userRepository.save(<%= instanceName %>.getUser());
}
```

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by mraible** [comment]
If you choose to use DTOs, doesn't the UserDTO contain all user's data by default?

**Comment by Falydoor** [comment]
It does but the entity DTO that has a relationship with User doesn't use UserDTO (it uses just userId).

**Comment by jdubois** [comment]
@Falydoor this is an excellent clean up, much easier to read and maintain, but it's quite hard to test. As you pass the whole test suite I guess you have no errors there, so I'm merging this (as anyway I couldn't do better tests manually).

**Comment by jdubois** [comment]
Whooops sorry I just noticed the [WIP] tag - @Falydoor did I do a mistake in merging this????

**Comment by Falydoor** [comment]
@jdubois The initial issue is fixed so I think it's fine to merge it!

**Comment by jsm174** [comment]
I just went to test this. Was able to successfully compile, but saving an entity in the gateway fails:

```
microservice2-app_1         | 2018-10-15 14:00:29.715 ERROR 1 --- [  XNIO-2 task-6] o.h.i.ExceptionMapperStandardImpl        : HHH000346: Error during managed flush [org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User]
microservice2-app_1         | 2018-10-15 14:00:29.720 ERROR 1 --- [  XNIO-2 task-6] o.z.problem.spring.common.AdviceTrait    : Internal Server Error
microservice2-app_1         |
microservice2-app_1         | org.springframework.dao.InvalidDataAccessApiUsageException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User; nested exception is java.lang.IllegalStateException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User
microservice2-app_1         | 	at org.springframework.orm.jpa.EntityManagerFactoryUtils.convertJpaAccessExceptionIfPossible(EntityManagerFactoryUtils.java:364)
microservice2-app_1         | 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:225)
microservice2-app_1         | 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:540)
microservice2-app_1         | 	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:746)

.
.
.

microservice2-app_1         | 	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
microservice2-app_1         | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
microservice2-app_1         | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
microservice2-app_1         | 	at java.lang.Thread.run(Thread.java:748)
microservice2-app_1         | Caused by: java.lang.IllegalStateException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User
microservice2-app_1         | 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:146)
microservice2-app_1         | 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:157)
microservice2-app_1         | 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:164)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1443)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:494)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3245)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2451)
microservice2-app_1         | 	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:473)
microservice2-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:156)
microservice2-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$100(JdbcResourceLocalTransactionCoordinatorImpl.java:38)
microservice2-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:231)
microservice2-app_1         | 	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:68)
microservice2-app_1         | 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:536)
microservice2-app_1         | 	... 131 common frames omitted
microservice2-app_1         | Caused by: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User
microservice2-app_1         | 	at org.hibernate.engine.spi.CascadingActions$8.noCascade(CascadingActions.java:380)
microservice2-app_1         | 	at org.hibernate.engine.internal.Cascade.cascade(Cascade.java:142)
microservice2-app_1         | 	at org.hibernate.event.internal.AbstractFlushingEventListener.cascadeOnFlush(AbstractFlushingEventListener.java:150)
microservice2-app_1         | 	at org.hibernate.event.internal.AbstractFlushingEventListener.prepareEntityFlushes(AbstractFlushingEventListener.java:141)
microservice2-app_1         | 	at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:74)
microservice2-app_1         | 	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:38)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1437)
microservice2-app_1         | 	... 140 common frames omitted
microservice2-app_1         |
microservice2-app_1         | 2018-10-15 14:00:29.770  WARN 1 --- [  XNIO-2 task-6] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [org.springframework.dao.InvalidDataAccessApiUsageException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User; nested exception is java.lang.IllegalStateException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User]
```

Here is the JDL:

```
application {
  config {
    baseName gateway,
    applicationType gateway,
    authenticationType oauth2,
    packageName com.example.gateway,
    prodDatabaseType postgresql,
    //clientFramework react,
    buildTool gradle
  }
  entities AA, B, C, D
}

application {
  config {
    baseName microservice1,
    applicationType microservice,
    authenticationType oauth2,
    packageName com.example.microservice1,
    prodDatabaseType postgresql,
    buildTool gradle
  }
  entities AA, B
}

application {
  config {
    baseName microservice2,
    applicationType microservice,
    authenticationType oauth2,
    packageName com.example.microservice2,
    prodDatabaseType postgresql,
    buildTool gradle
  }
  entities C, D
}

entity AA {
    data String
}

entity B {
    data String
}

entity C {
    data String
}

entity D {
    data String
}

relationship ManyToOne {
    D{user(login)} to User
}

dto * with mapstruct
service * with serviceClass

microservice AA, B with microservice1
microservice C, D with microservice2
```

screen shot 2018-10-15 at 10 09 02 am

**Comment by Falydoor** [comment]
I suggest you to open a new issue.
It is actually a little complicated because the microservice must save the User but the DTO doesn't contain all the data 😟.

File: generators/entity-server/templates/src/main/java/package/common/inject_template.ejs [hunk 1]
```diff
@@ -16,53 +16,31 @@
...
+        }
+    }
+    if (saveUserSnapshot && (viaService || constructorName.endsWith('Resource'))) {
+        beans.push({class: 'UserRepository', instance: 'userRepository'});
+    }
+_%>
 
-    private final UserRepository userRepository;
-<%_ } _%>
+    <%= beans.map(bean => `private ${bean.class} ${bean.instance};`).join('\n\n    ') %>
```
Comment (Akuka) [inline generators/entity-server/templates/src/main/java/package/common/inject_template.ejs]: Why you removed the 'final' ?

**Comment by Falydoor** [comment]
@Akuka my mistake, I forgot the keyword! See this PR #8601 for the fix.

**Pull request merged**